Default preferences center save success message

patrykgruszka opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
This PR adds success message in the default contact preferences page.

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Enable "Show contact preference settings" in Mautic configuration
4. Create contact, email and send it to generate unsubscribe link
5. Click unsubscribe link
6. Click "Save" in preferences page
7. Success message should appear on the page
![obraz](https://user-images.githubusercontent.com/8580942/189833069-7c23749f-f441-4b10-930a-e69a130d2a52.png)

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

patrykgruszka committed:
Contact preferences save success message

npracht commented:
@kuzmany wasn't it already existing in core or in a pending PR ?
<img width="409" alt="Capture d’écran 2022-09-13 à 09 24 27" src="https://user-images.githubusercontent.com/14075239/189837210-5be0df17-4f19-43eb-b428-cc49aa178dae.png">

kuzmany commented:
It was merged in 2.15 https://github.com/mautic/mautic/pull/6070

npracht commented:
@patrykgruszka can you explain what does it do new, this is unclear for me.

patrykgruszka commented:
Is it still available with the new builder? I can't find this option in the UI

npracht commented:
@patrykgruszka if you talk about grapesjs builder, unfortunately not, but there is this PR introducing all missing slots in GrapesJS builder : https://github.com/mautic/grapesjs-preset-mautic/pull/29

kuzmany commented:
@patrykgruszka ah I see. 
The slot is in the old builder. In the new builder  with this PR https://github.com/mautic/grapesjs-preset-mautic/pull/29

This PR added new option to Configuration to default preferences center (not landing page), am I right?

patrykgruszka commented:
@kuzmany, @npracht  yes this PR adds option to the default preferences center. I wasn't specific about this.

patrykgruszka committed:
Fix unit tests

patrykgruszka committed:
Change default preferences center success message label

patrykgruszka commented:
@npracht, @kuzmany I've updated the PR title, description and changed the textarea label to be more specific about the default preference center. Thanks for your feedback

patrykgruszka committed:
Merge branch '5.x' into gh-preferences-success-message

escopecz commented on app/bundles/EmailBundle/Form/Type/ConfigType.php:
```diff
@@ -74,6 +74,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    ? $options['data']['preferences_center_success_message']
+                    : $this->translator->trans(
+                        'mautic.email.preferences_center_success_message.text'
+                    ),
```
This could be simplified to
```suggestion
                'data'       => $options['data']['preferences_center_success_message'] ?? $this->translator->trans('mautic.email.preferences_center_success_message.text'),
```
But it will set the default message only if the config does not exist or is null (similar to `isset()`). Not sure if we need to cover also an empty string here.

escopecz commented on app/bundles/EmailBundle/Tests/Controller/PublicControllerFunctionalTest.php:
```diff
@@ -115,6 +133,15 @@ protected function getForm(?string $formTemplate): Form
         return $form;
     }
 
+    protected function getLead(): Lead
```
```suggestion
    protected function createLead(): Lead
```

escopecz commented on app/bundles/EmailBundle/Tests/Controller/PublicControllerFunctionalTest.php:
```diff
@@ -31,6 +34,20 @@ public function testUnsubscribeFormActionWithoutTheme(): void
+        $this->assertStringContainsString('/email/unsubscribe/tracking_hash_unsubscribe_form_email', $crawler->filter('form')->eq(0)->attr('action'));
+        $crawler = $this->client->submitForm('Save');
+
+        $this->assertEquals(1, $crawler->filter('#success-message-text')->count());
```
Can we assert also what the actual text is? That's the point of this feature, right?
```suggestion
        $this->assertEquals(1, $crawler->filter('#success-message-text')->count());
        $this->assertEquals('Success', $crawler->filter('#success-message-text')->text());
```

escopecz reviewed (commented):
The problem with this feature will be that the message won't be translate-able. The question is whether a custom success message feature is more important than translations.

Mautic actually can override translated strings quite easily. I found it in the code and wrote a documentation for it. Do you think this PR could be closed or simplified with translation overrides? 

https://github.com/mautic/mautic-documentation/pull/332

patrykgruszka committed:
Success message unit test enhancement

Co-authored-by: John Linhart <jan@linhart.email>

patrykgruszka committed:
Simplified preferences center save success message

patrykgruszka committed:
Merge branch '5.x' into gh-preferences-success-message

patrykgruszka committed:
Use translation instead harcdoded success message

patrykgruszka committed:
Merge branch 'gh-preferences-success-message' of github.com:patrykgruszka/mautic into gh-preferences-success-message

patrykgruszka commented:
@escopecz thanks for your suggestions. You're right, I haven't thought about translation of this message. Let's make it simpler, just with a translation mechanism. It should be easy to override with your instructions or even with a plugin.

escopecz commented on app/bundles/EmailBundle/Tests/Controller/PublicControllerFunctionalTest.php:
```diff
@@ -8,12 +8,13 @@
     protected function setUp(): void
     {
-        $this->configParams['show_contact_preferences'] = 1;
+        $this->configParams['show_contact_preferences']           = 1;
```
Can we revert this?
```suggestion
        $this->configParams['show_contact_preferences'] = 1;
```

escopecz reviewed (commented):
The code looks great. Thanks for the simplification! I have one more code style request.

patrykgruszka committed:
Code style fix

Co-authored-by: John Linhart <jan@linhart.email>

patrykgruszka commented:
@escopecz  sure, I've committed your suggestion

escopecz reviewed (approved):
The code looks good. Thank you!

jos0405 reviewed (approved):
Yes, the success message shows. Good enhancement.
![Screenshot 2022-10-07 at 15 43 01](https://user-images.githubusercontent.com/4246909/194572411-3e2003c1-9230-49dc-9f33-77788893a19c.png)

escopecz committed:
Merge branch '5.x' into gh-preferences-success-message

escopecz committed:
Merge branch '5.x' into gh-preferences-success-message
